sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala

```diff
@@ -342,9 +342,10 @@ object TypeCoercion {
         Intersect(newChildren.head, newChildren.last, isAll)
 
       case s: Union if s.childrenResolved &&
-          s.children.forall(_.output.length == s.children.head.output.length) && !s.resolved =>
+          s.children.forall(_.output.length == s.children.head.output.length) && !s.resolved &&
+          !s.byName =>
```

Comment:
nit: Since `s.byName` is constant, how about evaluating it first?
```
      case s @ Union(_, false, _) if s.childrenResolved && !s.byName &&
          s.children.forall(_.output.length == s.children.head.output.length) && !s.resolved =>
```